Content.Client/DragDrop/DragDropSystem.cs

```diff
@@ -18,479 +16,525 @@
             {
-                player.PopupMessage(Loc.GetString("drag-drop-system-out-of-range-text"));
+                continue;
             }
 
-            _dragDropHelper.EndDrag();
-            return false;
+            // highlight depending on whether its in or out of range
+            inRangeSprite.PostShader = valid.Value ? _dropTargetInRangeShader : _dropTargetOutOfRangeShader;
+            inRangeSprite.RenderOrder = EntityManager.CurrentTick.Value;
+            _highlightedSprites.Add(inRangeSprite);
         }
+    }
 
-        // TODO make this just use TargetOutlineSystem
-        private void HighlightTargets()
+    private void RemoveHighlights()
+    {
+        foreach (var highlightedSprite in _highlightedSprites)
         {
-            if (_dragDropHelper.Dragged == default || Deleted(_dragDropHelper.Dragged) ||
-                _dragShadow == default || Deleted(_dragShadow))
-            {
-                Logger.Warning("Programming error. Can't highlight drag and drop targets, not currently " +
-                               "dragging anything or dragged entity / shadow was deleted.");
-                return;
-            }
+            if (highlightedSprite.PostShader != _dropTargetInRangeShader && highlightedSprite.PostShader != _dropTargetOutOfRangeShader)
+                continue;
 
-            // highlights the possible targets which are visible
-            // and able to be dropped on by the current dragged entity
-
-            // remove current highlights
-            RemoveHighlights();
+            highlightedSprite.PostShader = null;
+            highlightedSprite.RenderOrder = 0;
+        }
 
-            // find possible targets on screen even if not reachable
-            // TODO: Duplicated in SpriteSystem and TargetOutlineSystem. Should probably be cached somewhere for a frame?
-            var mousePos = _eyeManager.ScreenToMap(_inputManager.MouseScreenPosition).Position;
-            var bounds = new Box2(mousePos - 1.5f, mousePos + 1.5f);
-            var pvsEntities = _lookup.GetEntitiesIntersecting(_eyeManager.CurrentMap, bounds, LookupFlags.Approximate | LookupFlags.Anchored);
-            foreach (var pvsEntity in pvsEntities)
-            {
-                if (!EntityManager.TryGetComponent(pvsEntity, out SpriteComponent? inRangeSprite) ||
-                    !inRangeSprite.Visible ||
-                    pvsEntity == _dragDropHelper.Dragged) continue;
+        _highlightedSprites.Clear();
+    }
 
-                // check if it's able to be dropped on by current dragged entity
-                var dropArgs = new DragDropEvent(_dragger, EntityManager.GetComponent<TransformComponent>(pvsEntity).Coordinates, _dragDropHelper.Dragged, pvsEntity);
+    /// <summary>
+    ///     Are these args valid for drag-drop?
+    /// </summary>
+    /// <returns>
+    /// Returns null if no interactions are available. Returns false if interactions exist but are not available currently.
```

Comment:
Description is misleading, because it also returns false when action blocker checks fail